src/internal/pfsdb/repos.go

```diff
@@ -0,0 +1,363 @@
+	}
+	var err error
+	if iter.index >= len(iter.repos) {
+		iter.index = 0
+		iter.offset += iter.limit
+		iter.repos, err = listRepoPage(ctx, iter.tx, iter.limit, iter.offset, iter.where, iter.whereVal)
+		if err != nil {
+			return errors.Wrap(err, "list repo page")
+		}
+		if len(iter.repos) == 0 {
+			return stream.EOS()
+		}
+	}
+	row := iter.repos[iter.index]
+	*dst, err = getRepoFromRepoRow(&row)
+	if id != nil {
+		*id = pachsql.ID(row.ID)
+	}
+	if err != nil {
+		return errors.Wrap(err, "getting repoInfo from repo row")
+	}
+	iter.index++
+	return nil
+}
+
+// Next advances the iterator by one row. It returns a stream.EOS when there are no more entries.
+func (iter *RepoIterator) Next(ctx context.Context, dst **pfs.RepoInfo) error {
+	return iter.NextWithID(ctx, nil, dst)
+}
+
+// ListRepo returns a RepoIterator that exposes a Next() function for retrieving *pfs.RepoInfo references.
+func ListRepo(ctx context.Context, tx *pachsql.Tx) (*RepoIterator, error) {
+	return listRepo(ctx, tx, "", nil)
+}
+
+// ListRepoByIdxType is like ListRepo but only iterates over repo.type = repoType.
+func ListRepoByIdxType(ctx context.Context, tx *pachsql.Tx, repoType string) (*RepoIterator, error) {
+	return listRepo(ctx, tx, "type", repoType)
+}
+
+// ListRepoByIdxName is like ListRepo but only iterates over repo.name = name.
+func ListRepoByIdxName(ctx context.Context, tx *pachsql.Tx, repoName string) (*RepoIterator, error) {
+	return listRepo(ctx, tx, "name", repoName)
+}
+
+// ListRepo returns a RepoIterator that exposes a Next() function for retrieving *pfs.RepoInfo references.
+func listRepo(ctx context.Context, tx *pachsql.Tx, where string, whereVal interface{}) (*RepoIterator, error) {
+	limit := 100
+	page, err := listRepoPage(ctx, tx, limit, 0, where, whereVal)
+	if err != nil {
+		return nil, errors.Wrap(err, "list repos")
+	}
+	iter := &RepoIterator{
+		repos: page,
+		limit: limit,
+		tx:    tx,
+	}
+	if where != "" && whereVal != nil {
+		iter.where = where
+		iter.whereVal = whereVal
+	}
+	return iter, nil
+}
+
+func listRepoPage(ctx context.Context, tx *pachsql.Tx, limit, offset int, where string, whereVal interface{}) ([]repoRow, error) {
+	var page []repoRow
+	if where != "" && whereVal != nil {
+		if err := tx.SelectContext(ctx, &page,
+			fmt.Sprintf("%s WHERE repo.%s = $1 GROUP BY repo.id, project.name ORDER BY repo.id ASC LIMIT $2 OFFSET $3;", getRepoAndBranches, where),
+			whereVal, limit, offset); err != nil {
+			return nil, errors.Wrap(err, "could not get repo page")
+		}
+		return page, nil
+	}
+	err := tx.SelectContext(ctx, &page, getRepoAndBranches+"GROUP BY repo.id, project.name ORDER BY repo.id ASC LIMIT $1 OFFSET $2 ;", limit, offset)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get repo page")
+	}
+	return page, nil
+}
+
+// CreateRepo creates an entry in the pfs.repos table.
+func CreateRepo(ctx context.Context, tx *pachsql.Tx, repo *pfs.RepoInfo) error {
```

Comment:
Should this return the `RepoID`?

Reply:
CreateProject does not either, but maybe it should 🤔